File: java/src/main/java/HandshakeCalculator.java
import java.util.Collections;
import java.util.List;
import java.util.ArrayList;

class HandshakeCalculator {

    List<Signal> calculateHandshake(final int number) {
        final List<Signal> result = new ArrayList<>();
        for (final Signal signal: Signal.values()) {
            if (((number >> signal.ordinal()) & 1) == 1) result.add(signal);
        }
        if ((number & 16) == 16) Collections.reverse(result);
        return result;
    }

}

File: kotlin/src/main/kotlin/HandshakeCalculator.kt
object HandshakeCalculator {
    private infix fun Int.isBitSet(bit: Int) = this and bit == bit

    fun calculateHandshake(number: Int): List<Signal> {
        return mutableListOf<Signal>().apply {
            if (number.isBitSet(1)) {
                add(Signal.WINK);
            }
            if (number.isBitSet(2)) {
                add(Signal.DOUBLE_BLINK);
            }
            if (number.isBitSet(4)) {
                add(Signal.CLOSE_YOUR_EYES);
            }
            if (number.isBitSet(8)) {
                add(Signal.JUMP);
            }
            if (number.isBitSet(16)) {
                reverse();
            }
        }
    }

}
